Allow user to request additional derivatives in DerivativeParsedMaterial

dschwen opened the pull request:
Closes #18402

dschwen committed:
Use symbol names (#18402)

dschwen committed:
Move test objects into test app (#18402)

dschwen committed:
Add parameter (#18402)

dschwen committed:
Add tests (#18402)

dschwen committed:
Revert "Move test objects into test app (#18402)"

This reverts commit 84a3936ea898533f2137f90401b5ad796243f074.

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/785093/) on 7eac97c wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/18460/site)

This comment will be updated on new commits.

moosebuild commented on large_media:
```diff
@@ -1 +1 @@
-Subproject commit def87c9916e53ad08034f93db8fd82fb42578a85
+Subproject commit 901c9f1ba4e258fa7047a02812d3038f0dbc8e9e
```
**Caution!** This contains a submodule update

dschwen committed:
Accommodate GCC (#18402)

jiangwen84 commented on framework/include/actions/MaterialDerivativeTestAction.h:
```diff
@@ -47,5 +47,5 @@ class MaterialDerivativeTestAction : public Action, public DerivativeMaterialPro
 
   /// every derivative given by a list of variables to derive w.r.t
-  std::map<MaterialPropertyName, std::vector<VariableName>> _derivatives;
+  std::map<MaterialPropertyName, std::vector<SymbolName>> _derivatives;
```
update the comment? It is not limited to `variables` any more, correct?

dschwen replied:
In this action we only derive w.r.t. MOOSE variables

jiangwen84 commented on framework/include/kernels/MaterialDerivativeTestKernelBase.h:
```diff
@@ -68,10 +68,9 @@ MaterialDerivativeTestKernelBase<T>::validParams()
   params.addRequiredParam<MaterialPropertyName>(
       "material_property", "Name of material property for which derivatives are to be tested.");
   params.addRequiredCoupledVar("args", "List of variables the material property depends on");
```
Is it still required to provide `coupled var`?  Can the `DerivativeParsedMaterial` be used to only calculate the derivative w.r.t. non-variables symbols?

jiangwen84 commented on framework/include/materials/ParsedMaterialHelper.h:
```diff
@@ -83,7 +86,7 @@ class ParsedMaterialHelper : public FunctionMaterialBase<is_ad>, public Function
 
   /// variable names used in the expression (depends on the map_mode)
-  std::vector<std::string> _variable_names;
+  std::vector<SymbolName> _symbol_names;
```
update the comment?

jiangwen84 commented on framework/include/utils/FunctionMaterialPropertyDescriptor.h:
```diff
@@ -69,17 +69,17 @@ class FunctionMaterialPropertyDescriptor : public DerivativeMaterialPropertyName
    * derivative w.r.t. that variable using the D[x,a] syntax
    */
-  bool dependsOn(const std::string & var) const;
+  bool dependsOn(const SymbolName & symbol) const;
```
update the comment?

jiangwen84 commented on framework/src/materials/DerivativeParsedMaterialHelper.C:
```diff
@@ -103,17 +132,21 @@ DerivativeParsedMaterialHelperTempl<is_ad>::recurseDerivative(unsigned int var,
 
   // variable we are deriving w.r.t.
-  auto derivative_var = _variable_names[var];
+  auto derivative_var = _derivative_symbol_table[var];
```
update the comment? variable -> symbol?

jiangwen84 commented on framework/src/auxkernels/SelfAux.C:
```diff
@@ -19,13 +19,19 @@ SelfAux::validParams()
 
-SelfAux::SelfAux(const InputParameters & parameters) : AuxKernel(parameters) {}
+SelfAux::SelfAux(const InputParameters & parameters)
+  : AuxKernel(parameters), _v(isCoupled("v") ? coupledValue("v") : _u)
```
I have not used `SelfAux` before. For the original implementation, `_u` is the `auxvariable` name? This `AuxKernel` sets the `Auxvariable` to itself value? I guess I did not quite get when we need this.

dschwen replied:
TBH I don't know what this AuxKernel was designed for. I just extended it, because we didn't seem to have an AuxKernel that did what I needed and this one was closest. To clarify I needed to create a copy of a given AuxVariable with a different name to enable exodiffing between to files that only diffed on the name of one variable.

jiangwen84 commented on framework/src/materials/DerivativeParsedMaterialHelper.C:
```diff
@@ -38,6 +42,30 @@ template <bool is_ad>
+  for (unsigned i = 0; i < _nargs; ++i)
+    _derivative_symbol_table.push_back(i);
+
+  for (auto & ads : additional_derivative_symbols)
+  {
+    auto it = std::find(_symbol_names.begin(), _symbol_names.end(), ads);
```
do we need to handle the repeated symbols accidentally provided by users?

jiangwen84 reviewed (changes requested):
Looks good. Some minor comments.

dschwen committed:
Add tests, address comments (#18402)

dschwen commented:
Ok, thanks @jiangwen84 ! I addressed the comments and added a few more tests to check the error messages (I had also forgotten to add an entry in the `tests` spec for a test input I had already committed).

dschwen commented:
added the missing SQA details

dschwen commented:
Thanks @jiangwen84 , there is a real bug in there that I'm working on fixing. I'll push an update shortly.

dschwen committed:
Fix out of range access (#18402)
